### ISSUE TITLE
latest release on PWS doesn't work

### ISSUE THREAD

**Issue by jistok** [oetiker/sourcey-buildpack#1]
Looks like the latest changes on PWS (maybe diego?) broke sourcey - it doesn't look like any of the dependencies or perl itself is actually being installed.  In a fit of laziness was forced to use https://github.com/judofyr/perloku instead.

**Comment by amhuber** [comment]
I just tested the sample Mojolicious app on a CF deployment with Diego and got the same thing:

```
2016-03-21T10:32:13.30-0700 [APP/0]      ERR Can't locate Mojolicious/Lite.pm in @INC (you may need to install the Mojolicious::Lite module) (@INC contains: /etc/perl /usr/local/lib/perl/5.18.2 /usr/local/share/perl/5.18.2 /usr/lib/perl5 /usr/share/perl5 /usr/lib/perl/5.18 /usr/share/perl/5.18 /usr/local/lib/site_perl .) at ./app-simple.pl line 2.
2016-03-21T10:32:13.30-0700 [APP/0]      ERR BEGIN failed--compilation aborted at ./app-simple.pl line 2.
```

It does work fine when running on a DEA, just not a Diego node.  Do you need any additional data to help troubleshoot?

Aaron

**Comment by oetiker** [comment]
sorry I don't know about the systems you are trying this on ... set `SOURCEY_VERBOSE=1` in your `SourceyBuild.sh` to get more information about what is going on.

**Comment by oetiker** [comment]
@jistok in your fork you have altered sourcey ... your problem is not related to that, is it?

**Comment by amhuber** [comment]
Diego is the new replacement for the DEAs written in Go (DEA Go -> Diego).  They will very soon deprecate the Ruby based DEAs and Diego will be the only option for all Cloud Foundry deployments, and that is already the case at the public PWS is my understanding.  Otherwise this is just a typical CF deployment (v231).  I'll grab a full verbose log and post it.

**Comment by oetiker** [comment]
hmmm I have not used cloud foundry in a few months, so I have not been following the latest developments ... when I used it, I just got the cf command, written in go to push my app to cf ... 

or are you talking about something else ?

**Comment by amhuber** [comment]
Yes, I'm talking about the internals of Cloud Foundry that are used to run the apps inside a container once you've uploaded them.  They are moving to a new back-end and it doesn't like Sourcey.  See details on Diego here:

http://docs.cloudfoundry.org/concepts/diego/diego-architecture.html
https://docs.cloudfoundry.org/concepts/diego/dea-vs-diego.html
https://github.com/cloudfoundry-incubator/diego-design-notes/blob/master/migrating-to-diego.md#diego-deltas

Here is the full log from a test deployment:

https://gist.github.com/amhuber/85d3a4f15041000e5ecc

It looks like the compilation of Perl 5.22.1 worked fine, but as you can see from the very end the app is running using the Perl from the rootfs (5.18.1), so perhaps the environment variables didn't get set up correctly?

Aaron

**Comment by amhuber** [comment]
Looking over the differences notes this jumps out at me:

https://github.com/cloudfoundry-incubator/diego-design-notes/blob/master/migrating-to-diego.md#environment-variable-interpolation

I'm guessing that might be what's breaking your startup script:

https://github.com/oetiker/sourcey-buildpack/blob/master/bin/release#L9-L15

Would this be easy to fix?

Aaron

**Comment by oetiker** [comment]
what seems odd, is that the upload droplet is only 1.4k ... have the requirements for build packs changed in the new release ?

when I created sourcey I found that documentation was sparse at best, and I had to ssh into a running instance to understand what they were actually doing ...

**Comment by amhuber** [comment]
This is the first buildpack that I've seen that hasn't just worked on Diego so at least at a high level not much has changed.

Aaron

**Comment by oetiker** [comment]
before I created sourcey I found that other build packs often were monsters of code layers upon layers of stuff in different languages real abominations ...

can you point me to a bit of documentation describing the environment present at build and at runtime ... directories, environemnt variables ...

**Comment by amhuber** [comment]
I think as you found before the previous documentation is a bit sparse.  Let me post about this in the CF newsgroup and see if we can get someone a little closer to the Diego code to respond with any suggestions.

Aaron

**Comment by amhuber** [comment]
Posted at http://cf-dev.70369.x6.nabble.com/Sourcey-buildpack-not-working-on-Diego-td4275.html.

**Comment by oetiker** [comment]
I just looked over https://docs.cloudfoundry.org/buildpacks/custom.html but there nothing seems changed. 

I have tried to document my findings about the build and runtime environments in the README.md ... someone in the know might be quick to spot the difference. 

The directory layout at build time:

```
/home
   /vcap (aka $HOME)
      /sourcey (aka $PREFIX)

/tmp
   /sourcey.$$ (aka $WORK_DIR)
   /cache (aka $CACHE_DIR)

/tmp
   /stage (aka $STAGE_DIR)
      /app (aka $BUILD_DIR)
      /sourcey
```

and

The directory layout at runtime.

```
/home
   /vcap
      /app  (aka $HOME !!!)
      /sourcey
```

maybe this is not true anymore ... and `$HOME` is now pointing to `/home/vcap` as one would expect ...  or we should not be using `$HOME` but  `$1` or something else entirely in the `release` script.

``` bash
#!/bin/bash

BUILD_DIR=$1

SOURCEY=sourcey
# Don't be fooled by the examples. The config_vars key gets ignored on cloudfoundry.
# It was just added to create a semblance of heroku compatibility
P5L=
if [ ! -f $BUILD_DIR/$SOURCEY/bin/perl ]; then
   P5L=" PERL5LIB='$BUILD_DIR/$SOURCEY/lib/perl5'"
fi
cat <<YAML
---
default_process_types:
    web: PATH=$BUILD_DIR/$SOURCEY/bin:\$PATH$P5L ./SourceyStart.sh
YAML
```

**Comment by jvshahid** [comment]
It looks like Diego only copies the `build_dir`, i.e. the app directory and ignores the `sourcey` directory. This is the line in [the app lifecycle](https://github.com/cloudfoundry-incubator/buildpack_app_lifecycle/blob/585dcf6f5e43230dea9a15f162564b3bcbde49ca/buildpackrunner/runner.go#L114) that packages the droplet. I think the fix is to include `sourcey` inside the app directory.

**Comment by oetiker** [comment]
@jvshahid how does caching work in the 'new' world ?

**Comment by oetiker** [comment]
please test the new Diego branch https://github.com/oetiker/sourcey-buildpack/tree/Diego

**Comment by jvshahid** [comment]
@oetiker the second argument to the `compile` script is the cache directory. Anything that is there when the buildpack is finished compilin will be available the next time the script is invoked.

**Comment by oetiker** [comment]
@jvshahid ok that has not changed then ...  @amhuber can you please test.

**Comment by emalm** [comment]
Thanks, @oetiker, I think @jvshahid is right about the staging process on Diego capturing only the contents of the build directory. I was working on a fix last night, at https://github.com/ematpl/sourcey-buildpack/tree/diego-fix, and I had also found I needed to remove the line in `lib/SourceBuildLib.inc` that set `PREFIX`, as it seemed to clobber the value in the compile script. Once I did that, staging seemed to work fine on both DEAs and Diego (and the droplets were cross-compatible), but I hadn't verified all the caching scenarios yet.

Best,
Eric, CF Runtime Diego PM

**Comment by amhuber** [comment]
I tested using cf push amhtest01 -b https://github.com/oetiker/sourcey-buildpack#Diego and it's better but still not working:

```
2016-03-22T08:13:35.98-0700 [STG/0]      OUT Staging complete
2016-03-22T08:13:35.98-0700 [STG/0]      OUT Uploading droplet, build artifacts cache...
2016-03-22T08:13:35.98-0700 [STG/0]      OUT Uploading droplet...
2016-03-22T08:13:35.98-0700 [STG/0]      OUT Uploading build artifacts cache...
2016-03-22T08:13:37.29-0700 [STG/0]      OUT Uploaded build artifacts cache (13M)
2016-03-22T08:13:42.32-0700 [STG/0]      OUT Uploaded droplet (13M)
2016-03-22T08:13:42.32-0700 [STG/0]      OUT Uploading complete
2016-03-22T08:13:42.78-0700 [CELL/0]     OUT Creating container
2016-03-22T08:13:43.05-0700 [CELL/0]     OUT Successfully created container
2016-03-22T08:13:44.67-0700 [CELL/0]     OUT Starting health monitoring of container
2016-03-22T08:13:44.69-0700 [APP/0]      ERR Can't locate Mojolicious/Lite.pm in @INC (you may need to install the Mojolicious::Lite module) (@INC contains: /home/vcap/sourcey/lib/perl5/x86_64-linux-thread-multi /home/vcap/sourcey/lib/perl5 /home/vcap/sourcey/lib/perl/x86_64-linux-thread-multi /home/vcap/sourcey/lib/perl .) at ./app-simple.pl line 2.
2016-03-22T08:13:44.69-0700 [APP/0]      ERR BEGIN failed--compilation aborted at ./app-simple.pl line 2.
```

The droplet is now the expected size so it's picking up the files and it did run with the right Perl libraries.  Sounds like maybe that extra tweak that Eric made is necessary.  I won't bother testing his fork since he already did.  :-)

Aaron

**Comment by oetiker** [comment]
I have merged some ideas from @ematpl s patch ... please try again

**Comment by amhuber** [comment]
That worked, and I also tested on re-stage that it used the cache and didn't compile Perl again:

```
Creating container
Successfully created container
Downloading app package...
Downloaded app package (1.2K)
Downloading build artifacts cache...
Downloaded build artifacts cache (13M)
Staging...
       Restoring binaries from /tmp/cache
          1962 lines total
Exit status 0
Staging complete
Uploading droplet, build artifacts cache...
Uploading build artifacts cache...
Uploading droplet...
Uploaded build artifacts cache (13M)
Uploaded droplet (13M)
Uploading complete

1 of 1 instances running
```

Looks good to me.  Thanks to all for helping on this!

Aaron

**Comment by oetiker** [comment]
does it still work on the old setup ? if so, I will merge it :)

**Comment by amhuber** [comment]
I'll do one more test forcing an app to DEA to verify and reply shortly.

**Comment by amhuber** [comment]
Looks good on DEA as well for both compiling and restaging from cache, ship it.  :-)

**Comment by emalm** [comment]
Thanks for checking all those cases, @amhuber!

**Comment by oetiker** [comment]
merged -> #3
